mypy/errors.py

```diff
@@ -422,22 +432,46 @@ def format_messages(self, error_info: List[ErrorInfo]) -> List[str]:
                 s = '{}: {}: {}'.format(srcloc, severity, message)
             else:
                 s = message
-            if self.show_error_codes and code and severity != 'note':
+            # If we show source code snippet, the error code is shown under it
+            # (with a pointer to error location), instead of after the error message.
+            if self.show_error_codes and code and not add_snippets and severity != 'note':
                 # If note has an error code, it is related to a previous error. Avoid
                 # displaying duplicate error codes.
                 s = '{}  [{}]'.format(s, code.code)
             a.append(s)
+            if add_snippets:
+                if severity == 'error' and source_lines and line > 0:
+                    source_line, offset = trim_source_line(source_lines[line - 1],
+                                                           DEFAULT_COLUMNS, column, MINIMUM_WIDTH)
+                    a.append(' ' * DEFAULT_SOURCE_OFFSET + source_line)
+                    # Also append a marker pointing to the error start location.
+                    a.append(' ' * (DEFAULT_SOURCE_OFFSET + column - offset) + WIGGLY_LINE +
+                             ' [{}]'.format(code.code) if code else '')
         return a
 
     def file_messages(self, path: str) -> List[str]:
         """Return a string list of new error messages from a given file.
 
-        Use a form suitable for displaying to the user.
+        Use a form suitable for displaying to the user. If add_snippets is True,
+        then one must also pass the source code and Python version.
         """
         if path not in self.error_info_map:
             return []
         self.flushed_files.add(path)
-        return self.format_messages(self.error_info_map[path])
+        source_lines = None
+        if self.options and self.options.show_source_code:
+            if self.fscache:
+                try:
```

Comment:
This try statement could be provided as a callback to avoid a depending on `fscache`.